contracts/0.8.9/LidoOracleNew.sol

```diff
@@ -0,0 +1,571 @@
+ *
+ * The goal of the oracle is to inform other parts of the system about balances controlled by the
+ * DAO on the ETH 2.0 side. The balances can go up because of reward accumulation and can go down
+ * because of slashing.
+ *
+ * The timeline is divided into consecutive frames. Every oracle member may push its report once
+ * per frame. When the equal reports reach the configurable 'quorum' value, this frame is
+ * considered finalized and the resulting report is pushed to Lido.
+ *
+ * Not all frames may come to a quorum. Oracles may report only to the first epoch of the frame and
+ * only if no quorum is reached for this epoch yet.
+ */
+contract LidoOracleNew is CommitteeQuorum, AccessControlEnumerable, ReportEpochChecker {
+    using ERC165Checker for address;
+    using UnstructuredStorage for bytes32;
+
+    event AllowedBeaconBalanceAnnualRelativeIncreaseSet(uint256 value);
+    event AllowedBeaconBalanceRelativeDecreaseSet(uint256 value);
+    event BeaconReportReceiverSet(address callback);
+
+    event CommitteeMemberReported(
+        uint256 epochId,
+        uint256 beaconBalance,
+        uint256 beaconValidators,
+        address caller,
+        uint256 wcBufferedEther,
+        uint256[] requestIdToFinalizeUpTo,
+        uint256[] finalizationPooledEtherAmount,
+        uint256[] finalizationSharesAmount
+    );
+
+    event ConsensusReached(
+        uint256 epochId,
+        uint256 beaconBalance,
+        uint256 beaconValidators,
+        uint256 wcBufferedEther,
+        uint256[] requestIdToFinalizeUpTo,
+        uint256[] finalizationPooledEtherAmount,
+        uint256[] finalizationSharesAmount
+    );
+
+    event PostTotalShares(
+         uint256 postTotalPooledEther,
+         uint256 preTotalPooledEther,
+         uint256 timeElapsed,
+         uint256 totalShares
+    );
+
+    event ContractVersionSet(uint256 version);
+
+
+    struct MemberReport {
+        // Consensus info
+        uint256 epochId;
+        // CL values
+        uint256 beaconValidators;
+        uint64 beaconBalanceGwei;
+        address[] stakingModules;
+        uint256[] nodeOperatorsWithExitedValidators;
+        uint64[] exitedValidatorsNumbers;
+        // EL values
+        uint256 wcBufferedEther;
+        // decision
+        uint256 newDepositBufferWithdrawalsReserve;
+        uint256[] requestIdToFinalizeUpTo;
+        uint256[] finalizationPooledEtherAmount;
+        uint256[] finalizationSharesAmount;
+    }
+
+    /// ACL
+    bytes32 constant public MANAGE_MEMBERS_ROLE = keccak256("MANAGE_MEMBERS_ROLE");
+    bytes32 constant public MANAGE_QUORUM_ROLE = keccak256("MANAGE_QUORUM_ROLE");
+    bytes32 constant public SET_BEACON_SPEC_ROLE = keccak256("SET_BEACON_SPEC_ROLE");
+    bytes32 constant public SET_REPORT_BOUNDARIES_ROLE = keccak256("SET_REPORT_BOUNDARIES_ROLE");
+    bytes32 constant public SET_BEACON_REPORT_RECEIVER_ROLE = keccak256("SET_BEACON_REPORT_RECEIVER_ROLE");
```

Comment:
want to introduce pre-rebase and post-rebase beacon report receivers (spare parts for the possible L2 staking design)